### ISSUE TITLE
QRNG Classifier Preprocessing Functions Improvements

### ISSUE THREAD

**Issue by sid-chava** [dorahacksglobal/qc-classifier#1]
We are building a classifier/prediction model that attempts to classify quantum computers based on quantum random number (QNRG) output data. We are looking to improve the performance (in other words, prediction accuracy) of our classifier model. We know that if we concatenate our binary QNRG sequences to increase the size of each training data line, the model is able to reach increasing accuracy, so we believe that it is possible for a model to distinguish between quantum computers with high accuracy - specifically because each QC has its unique quantum noise. 

In the main notebook in this repo, we have listed many preprocessing functions to try and improve the model's performance by adding various features to the training data, including many types of randomness extraction and statistical reduction functions such as entropy calculations. Please take a look at which preprocessing methods we have already defined and tested, and add any functions you can think of that might improve our classifier accuracy when applied to extract features from the input QNRG dataset (which is also included as a file in the repo). 

The benchmark and current model accuracy (with the optimized combination of features we have already tested before training our gradient booster model) is described in this repo's readme, so any improvements (to that score) will be useful and in the scope of the bounty! Please feel free to comment on this issue or email shrey@dorahacks.biz & sid@dorahacks.biz if you have any questions. 

You can duplicate the main jupyter notebook in this repo to play around with the preprocessing functions and model, and add your own methods.

**Comment by sid-chava** [comment]
Yes! That should not be an issue.

**Comment by ShreyK7** [comment]
Apologies, we just realized that the data file in the repo was a test data file with only 99 lines of QNRG output data. We have replaced that file with the full 14000 QNRG data file in the repo so you can test your models more robustly. Again, sorry for the confusion, the full-length data file is now in this repo. The labels at the end of each line classify which quantum computer the data was generated from (there are 4 devices we used).

**Comment by sid-chava** [comment]
Just pushed an update to fix that.

**Comment by ShreyK7** [comment]
@schance995 @0mWh reviewed your PR for this issue and you seem to have made really good improvements to the model -  great job! really like the sliding window you incorporated as well as the additional preprocessing extraction functions. 2 small things - a) could you try to isolate the accuracy improvements (for gradient boosting, random forest, and neural network models) from incorporating the sliding window method for generating additional data (unless I missed it in one of your notebooks). Would love to see how much that helped improve the accuracy. and b) we recently found out that the data labelled as the 1st quantum computer in our training QRNG dataset was in fact generated from a simulator and not an actual quantum device, so take out all the label 1 data from the original training data file and see how much the prediction accuracy improves with the remaining 3 actual quantum devices (it helped significantly when we tried it). Feel free to bundle these 2 things in the same PR as the updates you said you would make for more robust model testing, and let us know when you've made those updates. Finally, we are working to generate more QRNG testing data from different IBM and Microsoft Azure QCs, which should hopefully be ready and added to this repo by the weekend. Good work on this bounty again!

**Comment by 0mWh** [comment]
a) the helpfulness of the sliding window technique really depends on how we can determine the width of the quantum circuit's output bitstring -- is the circuit at the top of [process_logical_reduction.ipynb](https://github.com/dorahacksglobal/qc-classifier/pull/9/files#diff-4f124deae9f312002f077f32804c83cd58be711d777c92ac615f2c62b3e51bac) run through some other computation, or just output and appended?

b) that should definitely make it easier. for curiosity sake, was the output of (classical) device 1 tuned a similar noise model to that of (quantum) device 2, or something else?

**Comment by ShreyK7** [comment]
@0mWh I don't believe there was an noise tuning for device 1, the same CHSH circuit was just run through a classicial simulator instead of an actual QPU

**Comment by ShreyK7** [comment]
@0mWh as for your first question, I believe it's just raw output and appended for all the data after running the circuits (so in other words, the CHSH values), but i'll try to double check that soon - apologies for the confusion, the original data was generated by someone else on the team. Definitely nothing as complicated as a hashing algorithm though

**Comment by 0mWh** [comment]
everything has been adjusted and re-run without machine label 1, and now we are seeing approximately 75-77% accuracy with our models
